src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java

```diff
@@ -217,106 +219,142 @@ public StorageCalculator(boolean forArguments, boolean forVariadicFunction) {
+        ----------------+------------------+---------------------------------+-------------------------
+        Linux           | CW in regs       | CW on the stack                 | CW on the stack
+        MacOs           | CW in regs       | CW on the stack                 | CW on the stack
+        Windows, non-VF | CW in regs       | CW on the stack                 | CW on the stack
+        Windows, VF     | CW in regs       | CW split between regs and stack | CW on the stack
+         */
+        StructStorage[] structStorages(GroupLayout layout, boolean forHFA) {
+            int regType = forHFA ? StorageType.VECTOR : StorageType.INTEGER;
+            int numChunks = (int)Utils.alignUp(layout.byteSize(), MAX_COPY_SIZE) / MAX_COPY_SIZE;
+            int requiredStorages = forHFA ? layout.memberLayouts().size() : numChunks;
+            boolean hasEnoughRegisters = hasEnoughRegisters(regType, requiredStorages);
+
+            // For the ABI variants that pack arguments spilled to the
+            // stack, HFA arguments are spilled as if their individual
+            // fields had been allocated separately rather than as if the
+            // struct had been spilled as a whole.
+            boolean useFieldWiseSpill = requiresSubSlotStackPacking() && !forVarArgs;
+            boolean isFieldWise = forHFA && (hasEnoughRegisters || useFieldWiseSpill);
+            if (!isFieldWise) {
+                requiredStorages = numChunks;
+            }
+
+            boolean spillPartially = forVariadicFunction && spillsVariadicStructsPartially();
+            boolean furtherAllocationFromTheStack = !hasEnoughRegisters && !spillPartially;
+            if (furtherAllocationFromTheStack) {
                 // Any further allocations for this register type must
                 // be from the stack.
-                nRegs[type] = MAX_REGISTER_ARGUMENTS;
-                return null;
+                nRegs[regType] = MAX_REGISTER_ARGUMENTS;
             }
-        }
 
-        VMStorage[] regAlloc(int type, MemoryLayout layout) {
-            boolean spillRegistersPartially = forVariadicFunction && spillsVariadicStructsPartially();
+            if (requiresSubSlotStackPacking() && !isFieldWise) {
+                // Pad to the next stack slot boundary instead of packing
+                // additional arguments into the unused space.
+                alignStack(STACK_SLOT_SIZE);
+            }
 
-            return spillRegistersPartially ?
-                regAllocPartial(type, layout) :
-                regAlloc(type, requiredRegisters(layout));
-        }
+            StructStorage[] structStorages = new StructStorage[requiredStorages];
+            long offset = 0;
+            for (int i = 0; i < structStorages.length; i++) {
+                ValueLayout copyLayout;
+                if (isFieldWise) {
+                    // fields only, no padding
+                    copyLayout = (ValueLayout) layout.memberLayouts().get(i);
```

Comment:
Doesn't this contradict the comment? E.g. if we only copy fields, then we should skip padding layouts here (as they can be members of a group layout) ?

Reply:
The cast would fail if we ran into a padding layout, so there is an implicit assumption that we will not see any padding. I added the comment there to emphasize that. If we could have padding at this point, the offset computation we do later would also be incorrect.

Reply:
I've attempted to clarify the comment a bit.

Reply:
> I've attempted to clarify the comment a bit.

Just to check - where do we check that these structs do not have any padding inside? Does a meaningful exception get produced if that is not the case?

Reply:
Only HFAs are handled in this field-wise manner (see the table in the comment on this method as well). The assumption is that any struct with padding doesn't get classified as an HFA, and we can not get here (see [`jdk.internal.foreign.abi.aarch64.TypeClass::isHomogeneousFloatAggregate`](https://github.com/openjdk/panama-foreign/blob/08225e4f5e00837e974098a933f27267f3c0dbc5/src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/TypeClass.java#L61-L91). But, this is a non-local assumption, so the comment seems useful.

Reply:
Attempted to clarify the comment once more.

Reply:
Ok, the main issue was that by the time I got there I lost track of "isFieldWise implies HFA"